### DISCUSSION TITLE
Develop API 5 - Multi type queries with advanced_search - best practice?

### DISCUSSION THREAD

**Discussion by Andrew-McGee** [ampache/ampache#2877] [Q&A]
Hi folks.

As part of my frontend I am making a multi-search box with similar function to the one in the Ampache gui - which I should say is really cool.

I have the advanced_search API request working fine, even though it was very difficult* to understand from the docs! But at the moment I want my search results to return artists, albums and tracks from a single lookup and to do this I am really doing 3 separate API calls that look like this:

`action=advanced_search&type=artist&limit=5&rule_1=title&rule_1_operator=0&rule_1_input=$query`

`action=advanced_search&type=album&limit=5&rule_1=title&rule_1_operator=0&rule_1_input=$query`

`action=advanced_search&type=song&limit=5&rule_1=title&rule_1_operator=0&rule_1_input=$query`

Then I format the results on to a single page. It's still pretty quick because I use the limit parm.

But my question: Is there a better way to get multiple results back using a single query? I know there is the 'anywhere' rule but I couldn't understand exactly what this does and the docs say it only applies to the song items. It seems I have to use the type parm to specify the search boundaries.

Any idea or suggestion for optimal way to do this most appreciated.

*I will help to improve the docs with some of my notes when I get time. :)

**Comment by lachlan-00** [comment]
i'll have to have a deeper look.

the query is the same for all 3?

**Reply by Andrew-McGee** [reply]
One of them type=artist; the next type=album; then type=song. That gives me the 3 different search results.

**Reply by lachlan-00** [reply]
no i mean the `$query` parameter is the same and you want results for all 3 types in a single response?

I'm going to play on this one tomorrow

**Reply by Andrew-McGee** [reply]
Oh yes. Correct - same query for each. I mean the way I have implemented it works ok.  I just wondered if I was missing something and there was  a more efficient API call that returned all of these in one hit.

**Reply by lachlan-00** [reply]
just looking in the method and the search types being joined would be a lot of work to achieve this when it's possible anyway by changing the type.

probably best to keep it as you're doing without complicating it.

**Reply by Andrew-McGee** [reply]
ok makes sense. Thanks for checking.
